pkg/sync/sync.go

```diff
@@ -767,3 +776,106 @@ func (t *Task) RebuildStatus() (map[string]*ReplicaRebuildStatus, error) {
 
 	return replicaStatusMap, nil
 }
+
+func CloneSnapshot(engineControllerClient, fromControllerClient *client.ControllerClient, snapshotFileName string, exportBackingImageIfExist bool) error {
+	replicas, err := fromControllerClient.ReplicaList()
+	if err != nil {
+		return err
+	}
+	var sourceReplica *types.ControllerReplicaInfo
+	for _, r := range replicas {
+		if r.Mode == types.RW {
+			sourceReplica = r
+			break
+		}
+	}
+	if sourceReplica == nil {
+		return fmt.Errorf("cannot find a RW replica in the source volume for clonning")
+	}
+
+	replicas, err = engineControllerClient.ReplicaList()
+	if err != nil {
+		return err
+	}
+	for _, r := range replicas {
+		if r.Mode != types.RW {
+			return fmt.Errorf("cannot do snapshot clone because replica %v in %v mode. "+
+				"All replicas in the target volume must be in RW mode", r.Address, r.Mode)
+		}
+	}
+
+	taskErr := NewTaskError()
+	syncErrorMap := sync.Map{}
+	var wg sync.WaitGroup
+	wg.Add(len(replicas))
+
+	for _, r := range replicas {
+		go func(r *types.ControllerReplicaInfo) {
+			defer wg.Done()
+			repClient, err := replicaClient.NewReplicaClient(r.Address)
+			if err != nil {
+				syncErrorMap.Store(r.Address, err)
+				return
+			}
+			defer repClient.Close()
+			if err := repClient.CloneSnapshot(sourceReplica.Address, snapshotFileName, exportBackingImageIfExist); err != nil {
```

Comment:
This means when multiple replicas of the dest volume will only clone from the same source replica of the source volume. 

My 2 cents, in the future, probably think about how to share load to better leverage available replicas to clone instead of only one source replica.

Reply:
@innobead  Thanks for the advice! 

Currently we only allow the target volume to have exactly 1 replica while cloning is happening. After that, we scale up the replica. So, the multiple cloning is not an issue here. 

However, in the future if we want to speed up process, then yeah we can follow your advice.

Reply:
Thank for the explanation :+1: